### PR TITLE
Flash Decode GQA (and MQA) Improvements (Round 1)

### PR DESCRIPTION
**Pull request by caixunshiren**
### Ticket
This PR contains the round 1 improvements outlined in #12330 :

- [x] Support transpose_q, which both q of shape [1 x qh x b x h] and [1 x b x qh x h] are supported as input
- [x] Support GQA on a shared cache, which KV of shape [1 x kh x s x h] is supported
- [x] Add support for tensor indices for GQA

FYI @sraizada-tt @cglagovichTT 

Post commit pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/10956618476/job/30423043917

### CODE REVIEW

**Review by cfjchu — APPROVED** [review]

**Review by sraizada-tt — APPROVED** [review]

**Comment by caixunshiren** [comment]
all post commit: https://github.com/tenstorrent/tt-metal/actions/runs/10924399158

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.hpp [hunk 1]
```diff
@@ -18,6 +18,9 @@ struct ExecuteScaledDotProductAttentionGQADecode {
         const ttnn::Tensor &input_tensor_k,
         const ttnn::Tensor &input_tensor_v,
         const std::vector<uint32_t> cur_pos,
+        const std::optional<const Tensor> cur_pos_tensor = std::nullopt,
+        std::optional<bool> transpose_q = true,
+        std::optional<bool> share_cache = false,
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.hpp:23]: Should these just be regular args with default values?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comment

**Review by cglagovichTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.hpp (hunk 1 above)
Comment (caixunshiren) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.hpp:23]: Good point! My next pr actually changed it to regular args with default values.

**Review by uaydonat — APPROVED** [review]

**Pull request merged**